### ISSUE TITLE
[FEATURE REQUEST]: Virtual Switches for Modes

### ISSUE THREAD

**Issue by barneyonline** [MiguelRipoll23/homebridge-securitysystem#28]
(not an original idea) Create virtual switches for the modes (that are not disabled) so that the security system modes can be changed automatically based on user location.

Example use cases:
-First Person Arrives Home -> Away Mode Switch turned Off -> Security System mode set to Off
-Last Person Leaves Home -> Away Mode Switch turned On -> Security System mode set to Away

Obviously this is a workaround for Apple's automation rules but it would open up some more use cases. I'd recommend that this feature a) is optional and b) aligns with Disabled Modes

**Comment by macjl** [comment]
Hello,

For my own use, I have implemented same mechanism with an external plug-in (homebridge-http-switch, but can be an other), that use the web server to set and check the mode. I can share my configuration if you want.

But if the feature is implemented directly in this plug-in, I may use it too.

**Comment by MiguelRipoll23** [comment]
I discourage bypassing the HomeKit confirmation to use location triggers, even in the best conditions they can arm/disarm the security system unexpectedly at least one time.

This is something I experienced myself, it works flawless for months until one day you ask yourself why the lights faded out without you doing anything 😅 

This could be risky if the security system disarms during sleep time.

**Comment by barneyonline** [comment]
@macjl I also use an external plugin (homebridge-ifttt) as I'm controlling an Arlo system but obviously requires a bit of Eve automation to sync up correctly. Since homebridge-securitysystem already supports webhooks it would simplify the setup considerably. Could you share your config?

@MiguelRipoll23 I haven't experienced incorrect location triggering myself (except when I moved home and it took Google and Apple 3 days to update their Wi-Fi Location database). To reduce errors I use Hue Motion sensors and a timer switch (homebridge-automation-switches) to arm/disarm based on motion and "presence" switches. If you want to go further you could implement something like a Wi-Fi presence (homebridge-wifipresence) to improve the accuracy of triggers, but HomeKit location works well for me.

**Comment by macjl** [comment]
Here is the part of my configuration configuration I'm using for these switches :
```
   "accessories": [
      {
         "accessory":"HTTP-SWITCH",
         "name": "Arm away immediate",
         "switchType":"stateful",
         "onUrl": {
            "url":"http://localhost:9090/away?delay=false"
         },
         "offUrl": {
            "url":"http://localhost:9090/off"
         },
         "statusUrl": {
            "url":"http://localhost:9090/status"
         },
         "pullInterval": 2000,
         "statusPattern": "\"target_mode\":\"away\""
      },
      {
         "accessory":"HTTP-SWITCH",
         "name": "Arm away delayed",
         "switchType":"stateful",
         "onUrl": {
            "url":"http://localhost:9090/away?delay=true"
         },
         "offUrl": {
            "url":"http://localhost:9090/off"
         },
         "statusUrl": {
            "url":"http://localhost:9090/status"
         },
         "pullInterval": 2000,
         "statusPattern": "\"target_mode\":\"away\""
      },
      {
         "accessory":"HTTP-SWITCH",
         "name": "Arm night",
         "switchType":"stateful",
         "onUrl": {
            "url":"http://localhost:9090/night"
         },
         "offUrl": {
            "url":"http://localhost:9090/off"
         },
         "statusUrl": {
            "url":"http://localhost:9090/status"
         },
         "pullInterval": 2000,
         "statusPattern": "\"target_mode\":\"night\""
      }
   ]
```

**Comment by MiguelRipoll23** [comment]
You guys have convinced me.

This feature should be added with an option named `unsafe_mode_switches` so it’s clear that unless the user has a setup similar to @barneyonline it will be more unreliable and risky to use.

It’s quite amazing how you guys are making use of the features to expand the functionality of the plugin, after today’s stable update I’ll start the work for this and release a new beta as soon as posible so you can both simplify your setups.

Thanks!

**Comment by MiguelRipoll23** [comment]
Added to latest beta:

`npm i homebridge-securitysystem@2.6.0-beta.1`

**Comment by barneyonline** [comment]
That was quick @MiguelRipoll23 !
Testing now...

Minor feedback/personal-preference regarding the default switch names, could we append the security system "name" to the start? For example, switch name is "Arlo Mode Away" instead of "Mode Away"

Not an issue for Home app but makes it easier to identify in Eve
![IMG_4825](https://user-images.githubusercontent.com/38914183/80268017-95fa5600-86e7-11ea-947f-8f7a5e43883b.png)

**Comment by macjl** [comment]
Hello @barneyonline 
You can rename the switches in Home App. Then it show the name you want in Eve.

**Comment by macjl** [comment]
Hello @MiguelRipoll23,
I have a switch desynchronisation in last beta. The mode is Night, but all mode switches are off, and other strange behaviors.

**Comment by MiguelRipoll23** [comment]
@macjl are you on `v2.6.9-beta-2`, right?

This was happening to me too, I fixed some errors I did and after testing it was working well.

I’ll try again using several iOS devices to see if the state is actually updated or I’m being fooled by the Home app.

**Comment by macjl** [comment]
I have checked, and I’m on beta.1. I just specified the beta channel. Is the beta.2 right published?

**Comment by MiguelRipoll23** [comment]
My bad, it’s tagged as stable.

`npm i homebridge-securitysystem@2.6.0-beta.2`

I’m gonna fix the tag issue so it’s still part of the beta channel, thankfully that version is quite stable.

**Comment by macjl** [comment]
Ok. I have installed the Beta-2 from stable channel 😁

Just a question : is the « mode off » switch needed? For me, if all other mode switches are off, that you are in off mode. But maybe there are use case that needs it?

**Comment by macjl** [comment]
Found a limitation : I can’t switch off the « mode night » switch to disarm the system. Don’t tried with other mode from now, as it would result in triggering the system.

**Comment by MiguelRipoll23** [comment]
I can’t test this because we’re in lockdown but the Home app should ask you to confirm any mode change using the location trigger.

“When I arrive Home, turn off Security system” will ask for confirmation and that “off mode” switch will be a way to bypass it.

**Comment by MiguelRipoll23** [comment]
> Found a limitation : I can’t switch off the « mode night » switch to disarm the system. Don’t tried with other mode from now, as it would result in triggering the system. 

I’ll add an easy way to disarm the security system to all mode switches, that behavior will happen to all mode switches actually.

Thanks!

**Comment by MiguelRipoll23** [comment]
New beta update (v2.6.0-beta.3) released, read additional note at the bottom.

**Changelog**

✅ Added disarm functionality to mode switches.
✅ Added 'arm delay' characteristic to state saver.
✅ Fixed npm version tag to `beta`, removed `latest`.

**Known issues**
- When using the `save_state` option, mode switch corresponding to mode saved isn't on.
- When restarting Homebridge during arming, it gets stuck 'arming' on the Home app (affects previous stable versions).

**Additional notes**
When I published the first and second v2.6.0 beta it was late at night, there's a new characteristic available to control the arming delay on demand! I saw that @macjl was making use of the `delay` URL parameter so I decided to do something similar. It only affects mode switches but I'd like to make it global.

[Screenshot](https://i.imgur.com/FUU3MJ6.png)

**Comment by barneyonline** [comment]
Agree with @macjl, I also don't have a need for the "Mode Off" switch assuming the other mode switches can support an "off" state. Maybe @MiguelRipoll23 this should be removed or at least have an option to be removed?

@MiguelRipoll23 I've observed the following behaviour with v2.6.0-beta.3:

- Mode switch can not be turned to off state (displays "No Response" error), instead user needs to turn on the "Mode Off" switch which will then change other mode switches to "Off"
- Syncing between mode switches and security state work flawlessly
- Location-based rules (e.g. When Last Person Leaves, turn on "Mode Away" switch) work without a prompt from the Home app
- Logs record incorrect error "Missing webhook path for state." I can confirm the web hook is specified and works (sends command to IFTTT and arms Arlo).
[4/25/2020, 10:22:06 PM] [Arlo] Target state (Away)
[4/25/2020, 10:22:06 PM] [Arlo] Current state (Away)
[4/25/2020, 10:22:06 PM] [Arlo] Missing webhook path for state.
[4/25/2020, 10:22:07 PM] [Arlo] Webhook event (Sent)

**Comment by MiguelRipoll23** [comment]
> Maybe @MiguelRipoll23 this should be removed or at least have an option to be removed?

Sure, no problem!

> Syncing between mode switches and security state work flawlessly

Yup! Syncing works great now.

> Logs record incorrect error "Missing webhook path for state."

Could you share your plugin config here with sensitive data censored? I think that error is reporting that `webhook_target_away` is missing but I'm not sure.

—
I'm gonna test some under the hood improvements I did before releasing another beta today with the option to hide the mode off switch as I don't want to break anything 😄

**Comment by barneyonline** [comment]
Sure thing:

            "name": "Arlo",
            "default_mode": "Off",
            "disabled_modes": [
                "Home",
                "Night"
            ],
            "arm_seconds": 0,
            "trigger_seconds": 0,
            "siren_switch": false,
            "unsafe_mode_switches": true,
            "siren_mode_switches": false,
            "override_off": false,
            "save_state": true,
            "webhook_url": "https://maker.ifttt.com",
            "webhook_alert": "/trigger/Alarm_On/with/key/<REMOVED>",
            "webhook_target_away": "/trigger/Arlo_Arm/with/key/<REMOVED>",
            "webhook_off": "/trigger/Arlo_Disarm/with/key/<REMOVED>",
            "accessory": "security-system"

**Comment by MiguelRipoll23** [comment]
@barneyonline oh, I see what's happening:

1. Security system is arming to `Away` (target).
2. Sends webhook request as `webhook_target_away` is set.
3. Security system is armed to `Away` (current).
4. Reports _error_ as the option `webhook_current_away` is missing.

That message the way is implemented, it's more like a warning. So if you're wondering why the webhook requests aren't working because there's a typo or the option is missing, you can see that message and fix the plugin configuration.

**Things to improve**
- Type (current/target) needs to be added to the message.
- Make it clear that message isn't an error, something like:
  - `Webhook Path Current (None)`
  - `Webhook Path Target (None)`

Will that be better?

**Comment by MiguelRipoll23** [comment]
New beta update (v2.6.0-beta.4) released.

✅ Added option to hide `Mode Off` switch.
✅ Added Arming Delay characteristic to the security system to enable/disable delay.
✅ Fixed security system being stuck `Arming...` when restarting Homebridge while arming.
✅ Fixed mode switches not updating after loading saved state.
✅ Fixed security system's characteristic `Siren` not being reset when changing mode.
✅ Fixed error message when not setting webhook options for current/target states.

Hopefully everything sorted out 😊

**Comment by barneyonline** [comment]
@MiguelRipoll23 Ah yes I see what you mean.

I have tested v2.6.0-beta.5 and found no issues, works just as expected 😄 
Love your work @MiguelRipoll23

**Comment by MiguelRipoll23** [comment]
@barneyonline you guys keep my sanity during lockdown 😂

New stable update released (v2.6.0) with feature added, fixes a `No response` error when multiple sensor triggered events are canceled as well.

**Comment by barneyonline** [comment]
Glad I could help @MiguelRipoll23 😆
